### PR TITLE
test: enforce prompt-cache wire-prefix stability across all recorded cassettes

### PR DESCRIPTION
**Pull request by DouweM**
- Part of #6528 (slice C — the built-in invariant). Complements #6304's targeted UI-adapter round-trip tests.

## What

Every VCR-recorded test now doubles as a prompt-cache prefix regression guard. An autouse conftest fixture (modeled on `fail_partially_used_vcr_cassettes`) checks, at playback time, that the test's recorded conversation only ever *extends* the serialized provider-cache prefix (tools → system → messages, per provider shape). A framework change that silently starts moving wire prefixes — the #4338 class of bug, which costs users the full input price on every turn with no error — now fails the owning test with the exact divergent block and a pointer to the exemption.

Tests that move the prefix **deliberately** declare it on the test itself, and the reason is mandatory (a bare marker fails):

```python
@pytest.mark.moves_cache_prefix(reason='deliberate history compaction')
```

There are exactly ten such tests in the suite today: dynamic tool disclosure after `ToolSearch` discovery (tool-search evals, deferred-capability tests), OpenAI Responses history compaction, and one test that intentionally modifies history.

## Empirical basis

A prototype pass over the full corpus (1,177 cassettes, 386 with multi-request endpoints): **490 prefix-stable consecutive pairs, and zero unexplained wire-level busts** — the framework's request assembly is wire-prefix-stable today, and this locks that in. The only violating pairs were the ten deliberate movers above.

## How

- `tests/cassette_utils.py`: reusable helpers — `canonical_prefix_blocks(body, url)` (Anthropic Messages, OpenAI Chat/Responses, Google, Bedrock Converse shapes; single-block fields like plain-string system prompts and Google's `systemInstruction` Content dict serialize whole), `classify_prefix_pair` (extension / identical / shrunk / new-conversation / different-conversation / `{level}`-divergent), `iter_cassette_prefix_violations(path)`, and `check_cache_prefix_stability(node, path)` (marker + reason enforcement + failure message; unit-testable so the failure paths are coverable).
- Pair classification distinguishes real busts from benign cassette patterns: fresh conversations and independent conversations in one cassette (different first non-system message) are skipped; `shrunk` and mid-conversation divergence fail.
- `tests/conftest.py`: the `moves_cache_prefix(reason)` marker registration and the autouse enforcement fixture. Playback-only (`RecordMode.NONE`): while recording, the on-disk cassette isn't final; CI runs playback, so every recorded test is enforced there.
- `tests/test_cache_prefix_invariant.py`: unit tests for the helpers (synthetic violating cassette, provider shapes incl. the Google `systemInstruction` dict, classification edge cases, marker exemption + reason enforcement, fixture guards).
- `tests/AGENTS.md` documents the invariant and the marker convention.

Deterministic tool-args serialization (the paired prevention from #6528) needed no changes — the corpus scan found zero arg-ordering drift; the invariant will catch any future regression.

## Known limitation

Because enforcement runs at each test's own teardown, cassettes belonging to tests that are skipped in a given environment aren't checked there; the all-extras CI job runs effectively everything, so coverage in CI is near-complete. A standalone corpus-scan job could be added later if a stronger net is wanted.

### Checklist

- [ ] Any **AI generated code** has been reviewed line-by-line by the human PR author, who stands by it.
- [x] No **breaking changes** in accordance with the [version policy](https://github.com/pydantic/pydantic-ai/blob/main/docs/version-policy.md).
- [x] **PR title** is fit for the [release changelog](https://github.com/pydantic/pydantic-ai/releases).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/cassette_utils.py [hunk 1]
```diff
@@ -8,13 +8,176 @@
...
+        return 'openai-chat', blocks
+    if path.endswith('/responses') or '/responses' in path:
+        add('system', body.get('instructions'))
+        add('tools', body.get('tools'))
+        input_ = body.get('input')
+        add('messages', input_ if isinstance(input_, list) else [input_] if input_ is not None else None)
+        return 'openai-responses', blocks
+    if 'generativelanguage' in host or ':generateContent' in path or ':streamGenerateContent' in path:
+        add('tools', body.get('tools'))
+        add('system', body.get('systemInstruction') or body.get('system_instruction'))
```
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/cassette_utils.py:120]: Confirmed and fixed: `add()` now wraps any non-list value (plain-string system prompts and single Content dicts like Google's `systemInstruction`) as a single block instead of iterating it. Unit test added asserting the dict serializes whole and that a changed `systemInstruction` classifies as `system-divergent`. A fresh corpus scan with the fix shows no newly-visible violations — the recorded Google conversations keep their system instruction stable — so no new exemptions were needed.

File: tests/cassette_utils.py [hunk 2]
```diff
@@ -8,13 +8,196 @@
...
+    pair_index: int
+    level: str
+    block_index: int
+    earlier_block: str
+    later_block: str
+
+
+def check_cache_prefix_stability(node: pytest.Item, cassette_path: Path) -> None:
+    """Fail when a cassette moves its provider-cache wire prefix without an exemption."""
+    if node.get_closest_marker('moves_cache_prefix'):
```
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/cassette_utils.py]: Done — `check_cache_prefix_stability` now fails a bare `@pytest.mark.moves_cache_prefix` with no `reason=`, with a unit test covering it. (Note the allowlist this comment refers to has since been replaced by the marker per maintainer feedback, so the enforcement lives in the marker check.)

File: tests/cassette_utils.py [hunk 3]
```diff
@@ -8,13 +8,250 @@
...
+    Within a single run the agent loop only appends assistant and tool-result messages, never a new
+    user prompt, so a genuine user turn is the reliable boundary signal. Tool and function results are
+    carried on user-role messages by several providers (Anthropic/Bedrock `tool_result`/`toolResult`
+    content, Google `functionResponse` parts, OpenAI Responses `function_call_output`); those are part
+    of the same turn, so they are not counted as a new one.
+    """
+    try:
+        message = json.loads(block)
+    except Exception:
+        return False
```
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/cassette_utils.py:146]: Done in a8117c6a5 — narrowed to `json.JSONDecodeError` in `is_new_user_turn` and `(AttributeError, TypeError, json.JSONDecodeError)` in `conversation_identity` (its `try` wraps `.get('role')`, which raises `AttributeError` on a non-object parse — the `test_classify_prefix_pair_non_object_message_blocks` path). ruff `BLE001` clean and the 19 unit tests still pass.

File: tests/cassette_utils.py [hunk 4]
```diff
@@ -8,13 +8,250 @@
...
+    if level == 'messages' and divergent_index == first_message_index:
+        return 'new-conversation', divergent_index
+
+    def conversation_identity(blocks: list[PrefixBlock]) -> str | None:
+        for block_level, block in blocks:
+            if block_level != 'messages':
+                continue
+            try:
+                role = json.loads(block).get('role')
+            except Exception:
```
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/cassette_utils.py]: Done in a8117c6a5 — narrowed to `json.JSONDecodeError` in `is_new_user_turn` and `(AttributeError, TypeError, json.JSONDecodeError)` in `conversation_identity` (its `try` wraps `.get('role')`, which raises `AttributeError` on a non-object parse — the `test_classify_prefix_pair_non_object_message_blocks` path). ruff `BLE001` clean and the 19 unit tests still pass.

File: tests/cassette_utils.py [hunk 5]
```diff
@@ -8,13 +8,250 @@
...
+    divergent_index = next((i for i in range(shared_length) if a[i] != b[i]), shared_length)
+    if divergent_index == len(a) and len(b) > len(a):
+        return 'extension', -1
+    if divergent_index == len(b) and len(a) > len(b):
+        return 'shrunk', divergent_index
+
+    level = a[divergent_index][0] if divergent_index < len(a) else b[divergent_index][0]
+    first_message_index = next((i for i, (block_level, _) in enumerate(a) if block_level == 'messages'), None)
+    if level == 'messages' and divergent_index == first_message_index:
+        return 'new-conversation', divergent_index
```
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/cassette_utils.py]: Good catch — fixed in fc8f060b0, though the suggested patch wouldn't have covered the example (it diverges at index 1 while `1 < len(a)`, so `divergent_index >= len(a)` is false and `level` stays `messages`). Root cause is that `level` came from `a[divergent_index]` alone. Since both diverging blocks always exist past the extension/shrunk guards, I now derive the level from whichever changed block is earliest in cache order (an inserted `tools` block is the real change, not the `messages` block it shifted back) and only classify `new-conversation` when *both* sides diverge within the message history. Your example now returns `('tools-divergent', 1)`; added it as a regression test. Re-scanned all 1,212 cassettes — no previously-hidden movers surfaced.

File: tests/cassette_utils.py [hunk 6]
```diff
@@ -8,13 +8,262 @@
...
+    if path.endswith('/v1/messages') or (host == 'api.anthropic.com' and '/messages' in path):
+        add('tools', body.get('tools'))
+        add('system', body.get('system'))
+        add('messages', body.get('messages'))
+        return 'anthropic', blocks
+    if path.endswith('/chat/completions'):
+        add('tools', body.get('tools'))
+        add('messages', body.get('messages'))
+        return 'openai-chat', blocks
+    if path.endswith('/responses') or '/responses' in path:
```
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/cassette_utils.py]: Fixed in d59ed37cb, though not with the host guard — OpenAI Responses is also served from Azure and OpenRouter hosts (both in the corpus), so gating on `host == 'api.openai.com'` would wrongly drop them. Also, `path.endswith('/responses')` is already subsumed by `'/responses' in path`, so the check was purely the loose substring. Tightening to `path.endswith('/responses')` keeps every host's create endpoint and excludes the auxiliary sub-paths (`/responses/compact`, `/responses/input_tokens`) that were being pooled with it. That pooling turned out to be masking a false signal: with `/compact` no longer compared against the create body, the three `compact_*` tests are prefix-stable at the create endpoint, so I dropped their now-misleading `moves_cache_prefix` markers too.

File: tests/cassette_utils.py [hunk 7]
```diff
@@ -8,13 +8,266 @@
...
+        if items is None:
+            return
+        # Some fields hold a single block rather than a list: a plain-string system prompt
+        # (Anthropic/Bedrock), or Google's `systemInstruction`, which is one Content *dict* --
+        # iterating it would silently reduce it to its keys and blind the check to content changes.
+        block_items: list[Any] = items if _is_list(items) else [items]
+        for item in block_items:
+            blocks.append((level, json.dumps(item)))
+
+    if path.endswith('/v1/messages') or (host == 'api.anthropic.com' and '/messages' in path):
```
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/cassette_utils.py:99]: Fixed in f8df187bd, at the root rather than per-provider: the grouping key in `iter_cassette_prefix_violations` is now `(host, path, shape)` instead of `(host, shape)`. That stops pooling `/v1/messages/count_tokens`, `:countTokens`, and any sibling endpoint with the generation endpoint, and also covers a different model/deployment carried in the path (Bedrock `/model/{id}/converse`, Google `/models/{model}:generateContent`, Azure deployments) — which a per-provider suffix tweak wouldn't. Re-scanned all cassettes: same 8 marked movers, no new violators, so legitimate multi-turn pairs (same path throughout a run) are unaffected.
